File: problems/chapter16/miiitomi/007.cpp
// https://onlinejudge.u-aizu.ac.jp/services/room.html#ACPC2018Day1/problems/H

#include <iostream>
#include <vector>
using namespace std;

const int INF = 1e9;

struct Edge {
    int rev, to, cap;
    Edge(int r, int t, int c) : rev(r), to(t), cap(c) {}
};

int rec(vector<vector<Edge>> &G, vector<bool> &seen, int v, int t, int f) {
    if (seen[v]) return 0;
    seen[v] = true;
    if (v == t) return f;

    for (auto &e : G[v]) {
        if (e.cap == 0) continue;
        int flow = rec(G, seen, e.to, t, min(f, e.cap));
        if (flow == 0) continue;
        e.cap -= flow;
        G[e.to][e.rev].cap += flow;
        return flow;
    }
    return 0;
}

int FordFulkerson(vector<vector<Edge>> &G, int s, int t) {
    int ans = 0;
    while (true) {
        vector<bool> seen((int)G.size(), false);
        int flow = rec(G, seen, s, t, INF);
        if (flow == 0) return ans;
        else ans += flow;
    }
}

int main() {
    int R, C;
    cin >> R >> C;

    vector<pair<int, int>> V;
    for (int i = 0; i < R; i++) {
        string s;
        cin >> s;
        for (int j = 0; j < C; j++) {
            if (s[j] == '#') V.push_back(make_pair(i, j));
        }
    }

    int s = (int)V.size();
    int t = (int)V.size() + 1;

    int num_edges = 0;

    vector<vector<Edge>> G((int)V.size()+2);
    for (int i = 0; i < (int)V.size(); i++) {
        pair<int, int> n = V[i];
        int j;

        pair<int, int> n_shita = make_pair(n.first+1, n.second);
        auto iter_shita = lower_bound(V.begin(), V.end(), n_shita);
        if (iter_shita->first == n_shita.first && iter_shita->second == n_shita.second) {
            num_edges++;
            j = distance(V.begin(), iter_shita);
            int nv = (int)G.size();
            G.push_back((vector<Edge>){});
            G[i].push_back(Edge((int)G[nv].size(), nv, INF));
            G[nv].push_back(Edge((int)G[i].size()-1, i, 0));
            G[j].push_back(Edge((int)G[nv].size(), nv, INF));
            G[nv].push_back(Edge((int)G[j].size()-1, j, 0));
            G[nv].push_back(Edge((int)G[t].size(), t, 1));
            G[t].push_back(Edge((int)G[nv].size()-1, nv, 0));
        }

        pair<int, int> n_migi = make_pair(n.first, n.second+1);
        auto iter_migi = lower_bound(V.begin(), V.end(), n_migi);
        if (iter_migi->first == n_migi.first && iter_migi->second == n_migi.second) {
            num_edges++;
            j = distance(V.begin(), iter_migi);
            int nv = (int)G.size();
            G.push_back((vector<Edge>){});
            G[nv].push_back(Edge((int)G[i].size(), i, INF));
            G[i].push_back(Edge((int)G[nv].size()-1, nv, 0));
            G[nv].push_back(Edge((int)G[j].size(), j, INF));
            G[j].push_back(Edge((int)G[nv].size()-1, nv, 0));
            G[s].push_back(Edge((int)G[nv].size(), nv, 1));
            G[nv].push_back(Edge((int)G[s].size()-1, s, 0));
        }
    }

    int ff = FordFulkerson(G, s, t);
    cout << (int)V.size() - num_edges + ff << endl;
    return 0;
}

File: problems/chapter16/mshibatatt/007.py
# https://onlinejudge.u-aizu.ac.jp/services/room.html#ACPC2018Day1/problems/H
import sys
sys.setrecursionlimit(10000)

class FordFulkerson:
    def __init__(self, N):
        self.N = N
        self.G = [[] for i in range(N)]
 
    def add_edge(self, fr, to, cap):
        forward = [to, cap, None]
        forward[2] = backward = [fr, 0, forward]
        self.G[fr].append(forward)
        self.G[to].append(backward)
 
    def add_multi_edge(self, v1, v2, cap1, cap2):
        edge1 = [v2, cap1, None]
        edge1[2] = edge2 = [v1, cap2, edge1]
        self.G[v1].append(edge1)
        self.G[v2].append(edge2)
 
    def dfs(self, v, t, f):
        if v == t:
            return f
        seen = self.seen
        seen[v] = True
        for e in self.G[v]:
            w, cap, rev = e
            if cap and not seen[w]:
                d = self.dfs(w, t, min(f, cap))
                if d:
                    e[1] -= d
                    rev[1] += d
                    return d
        return 0
 
    def flow(self, s, t):
        flow = 0
        f = INF = float('inf')
        N = self.N
        while f:
            self.seen = [False]*N
            f = self.dfs(s, t, INF)
            flow += f
        return flow

def main():
    r, c = map(int, input().split())
    G = []
    for _ in range(r):
        G.append(input())

    # consider as minimun capacity cut
    ff = FordFulkerson(r*c*3 + 2)
    s, t = r*c*3, r*c*3 + 1 
    num_blank, num_adjacent = 0, 0
    for i in range(r):
        for j in range(c):
            if G[i][j] == '#':
                num_blank += 1
                v = i*c + j
                # if adjacent in row, cannot both 0
                if i > 0 and G[i-1][j] == '#':
                    num_adjacent += 1
                    newv = v + r*c
                    ff.add_edge(newv, t, 1)
                    ff.add_edge(v, newv, float('inf'))
                    ff.add_edge(v-c, newv, float('inf'))
                    
                # if adjacent in col, cannot both 1
                if j > 0 and G[i][j-1] == '#':
                    num_adjacent += 1
                    newv = v + r*c*2
                    ff.add_edge(s, newv, 1)
                    ff.add_edge(newv, v, float('inf'))
                    ff.add_edge(newv, v-1, float('inf'))
    
    output = num_blank - num_adjacent + ff.flow(s, t)
    print(output)

if __name__=='__main__':
    main()